mono/metadata/loader.c

```diff
@@ -1704,30 +1704,19 @@ pinvoke_probe_for_symbol (MonoDl *module, MonoMethodPInvoke *piinfo, const char
 	if (piinfo->piflags & PINVOKE_ATTRIBUTE_NO_MANGLE) {
 		error_msg = mono_dl_symbol (module, import, &addr); 
 	} else {
-		char *mangled_name = NULL, *mangled_name2 = NULL;
-		int mangle_charset;
-		int mangle_stdcall;
-		int mangle_param_count;
-#ifdef TARGET_WIN32
-		int param_count;
-#endif
-
 		/*
 		 * Search using a variety of mangled names
 		 */
-		for (mangle_charset = 0; mangle_charset <= 1; mangle_charset ++) {
-			for (mangle_stdcall = 0; mangle_stdcall <= 1; mangle_stdcall ++) {
-				gboolean need_param_count = FALSE;
+		for (int mangle_stdcall = 0; mangle_stdcall <= 1 && addr == NULL; mangle_stdcall++) {
 #ifdef TARGET_WIN32
-				if (mangle_stdcall > 0)
-					need_param_count = TRUE;
+			const int max_managle_param_count = (mangle_stdcall == 0) ? 0 : 256;
```

Comment:
Is there no signature with a param count, that we can just probe for once?
If so, great, use that.
If not, can you allocate the buffer just once instead of in a loop?

This mangling should be limited not just to Windows, but Windows/x86.
I'm pretty sure WIN32 means any architecture Windows, not just 32bit.

Reply:
Also would be good to put an `#if` around the loop bounds, or to terminate the second iteration right away.
i.e. be sure to iterate just once on all but Windows/x86.

Either like this:

```
#if win32 && x86
			for (mangle_stdcall = 0; mangle_stdcall <= 1; mangle_stdcall ++)
#else
			for (mangle_stdcall = 0; mangle_stdcall <= 0; mangle_stdcall ++)
#endif
{
```

or like this:

```
			for (mangle_stdcall = 0; mangle_stdcall <= 1; mangle_stdcall ++) { // no change this line 
#if !win32 || !x86
if (mangle_stdcall)
   break;
#endif 
```

Reply:
Tried to not change the probing behavior around param count from original logic, but I agree that it should probably be applied to just 32-bit Windows.

Reply:
I see you do use the signature for 0 params.
 But it looks like then the brute force path skips 0.

Reply:
Signature mangling is only used when doing the std_call path and was kept as it was implemented originally.

Reply:
I think it'd be best to start at -2 and have -2 and -1 both be special cases.
-2: unmangled; this will work the vast majority of the time, i.e. for the OS
-1: use the managed signature param count
>=0 the guessing that is present
And definitely this should be x86 only.
No other architecture, 32bit or 64bit, has multiple calling conventions or C mangling (ignoring amd64 vectorcall, granted).

Reply:
Oh maybe the stdcall check flows from the caller and it is skipped then for amd64. I'd have to step through it. :(

Reply:
Yes that implementation is a mess, so I didn't want to rewrite it to much :) I believe the Current change is solving the problem with unicode symbols and keep the old behavior intact.

Reply:
I think we should keep changes to minimum, but agree that we should disable the parameter probing on 64-bit Windows.

Reply:
Reduced parameter probing to Windows x86.